Applet/src/com/android/javacard/keymaster/KMKeymasterApplet.java

```diff
@@ -1096,6 +1096,25 @@ private void processComputeSharedHmacCmd(APDU apdu) {
     sendOutgoing(apdu);
   }
 
+  private byte validateUpgradeKeyForTag(short tag, short systemParam) {
+    // validate characteristics to be upgraded.
+    tmpVariables[0] = KMKeyParameters.findTag(KMType.UINT_TAG, tag, data[HW_PARAMETERS]);
+    tmpVariables[0] = KMIntegerTag.cast(tmpVariables[0]).getValue();
+    tmpVariables[1] = KMInteger.uint_8((byte) 0);
+    if (tmpVariables[0] != KMType.INVALID_VALUE) {
+      if ((tag == KMType.OS_VERSION
+              && KMInteger.compare(tmpVariables[0], systemParam) == 1
+              && KMInteger.compare(systemParam, tmpVariables[1]) == 0)
+              || (KMInteger.compare(tmpVariables[0], systemParam) == -1)) {
+        // Key needs upgrade.
+        return (byte) 1;
+      } else if (KMInteger.compare(tmpVariables[0], systemParam) == 1) {
+        KMException.throwIt(KMError.INVALID_ARGUMENT);
+      }
+    }
+    return (byte) 0;
```

Comment:
if you fail the else if and versions are valid then you will need to ruturn 1 and also if 1104 is false then you will need to return 0 (by deafult).

Reply:
Resolved.